### ISSUE TITLE
ClassCastException when declaring a constant Array binding and injecting a missing binding

### ISSUE THREAD

**Issue by venator85** [Kodein-Framework/Kodein-DI#149]
Hi, I'm using Kodein 5.2.0 Generic on Android (I only included `'org.kodein.di:kodein-di-generic-jvm:5.2.0'`).
With the following kodein instance, that defines a constant `ByteArray` binding:

    val kodein = Kodein {
        constant("foo") with "bar".toByteArray()
    }

if I try to inject a missing binding:

    class A(kodein: Kodein) {
        private val random: Random by kodein.instance()

        fun f() {
            random.nextInt()
        }
    }

I get this exception:

    E/AndroidRuntime: FATAL EXCEPTION: main
        Process: com.xxx, PID: 6395
        java.lang.ClassCastException: libcore.reflect.GenericArrayTypeImpl cannot be cast to java.lang.Class
            at org.kodein.di.ParameterizedTypeToken.getRawType(types.kt:177)
            at org.kodein.di.ParameterizedTypeToken.getRaw(types.kt:202)
            at org.kodein.di.TypeToken$DefaultImpls.isAssignableFrom(typeToken.kt:82)
            at org.kodein.di.JVMTypeToken.isAssignableFrom(types.kt:143)
            at org.kodein.di.ClassTypeToken.isAssignableFrom(types.kt:302)
            at org.kodein.di.internal.TypeChecker$Up.check(KodeinTreeImpl.kt:16)
            at org.kodein.di.internal.KodeinTreeImpl$findBySpecs$1.invoke(KodeinTreeImpl.kt:59)
            at org.kodein.di.internal.KodeinTreeImpl$findBySpecs$1.invoke(KodeinTreeImpl.kt:28)
            at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:110)
            at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
            at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:250)
            at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:242)
            at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:108)
            at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
            at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:250)
            at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:242)
            at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:108)
            at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
            at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:250)
            at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:242)
            at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:108)
            at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
            at kotlin.sequences.TransformingSequence$iterator$1.hasNext(Sequences.kt:153)
            at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:647)
            at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:677)
            at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:668)
            at org.kodein.di.internal.KodeinTreeImpl.findBySpecs(KodeinTreeImpl.kt:81)
            at org.kodein.di.internal.KodeinTreeImpl.find(KodeinTreeImpl.kt:99)
            at org.kodein.di.KodeinTree$DefaultImpls.find$default(KodeinTree.kt:32)
            at org.kodein.di.internal.KodeinContainerImpl.factory(KodeinContainerImpl.kt:146)
            at org.kodein.di.KodeinContainer$DefaultImpls.factory$default(KodeinContainer.kt:33)
            at org.kodein.di.KodeinContainer$DefaultImpls.provider(KodeinContainer.kt:80)
            at org.kodein.di.internal.KodeinContainerImpl.provider(KodeinContainerImpl.kt:7)
            at org.kodein.di.KodeinContainer$DefaultImpls.provider$default(KodeinContainer.kt:79)
            at org.kodein.di.KodeinAwareKt$Instance$1.invoke(KodeinAware.kt:152)
            at org.kodein.di.KodeinProperty$provideDelegate$1.invoke(properties.kt:39)
            at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)

Interestingly, with kodein 5.2.0 erased, I get a much more helpful message:

    org.kodein.di.Kodein$NotFoundException: No binding found for bind<Random>() with ? { ? }

If I remove the constant binding, I get the same helpful message in kodein generic too.

**Comment by SalomonBrys** [comment]
Thanks a lot for reporting.
This is due to a difference in array reflection in Android & in the JVM.

I've corrected this issue. It will be published in 5.2.1 & in 5.3.0.

I'll close this issue when they are released.

**Comment by chaschev** [comment]
Got the same exception for calling

```Kodein.direct.Instance(TT(configClass))```

Stacktrace:

```
Exception in thread "main" java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.TypeVariableImpl cannot be cast to java.lang.Class
	at org.kodein.di.ParameterizedTypeToken.getRawType(types.kt:177)
	at org.kodein.di.ParameterizedTypeToken.getRaw(types.kt:202)
	at org.kodein.di.TypeToken$DefaultImpls.isAssignableFrom(typeToken.kt:82)
	at org.kodein.di.JVMTypeToken.isAssignableFrom(types.kt:143)
	at org.kodein.di.TypeToken$DefaultImpls.isAssignableFrom(typeToken.kt:89)
	at org.kodein.di.JVMTypeToken.isAssignableFrom(types.kt:143)
	at org.kodein.di.TypeToken$DefaultImpls.isAssignableFrom(typeToken.kt:89)
	at org.kodein.di.JVMTypeToken.isAssignableFrom(types.kt:143)
	at org.kodein.di.ClassTypeToken.isAssignableFrom(types.kt:302)
	at org.kodein.di.internal.TypeChecker$Up.check(KodeinTreeImpl.kt:16)
	at org.kodein.di.internal.KodeinTreeImpl$findBySpecs$1.invoke(KodeinTreeImpl.kt:59)
	at org.kodein.di.internal.KodeinTreeImpl$findBySpecs$1.invoke(KodeinTreeImpl.kt:28)
	at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:110)
	at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
	at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:250)
	at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:242)
	at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:108)
	at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
	at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:250)
	at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:242)
	at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:108)
	at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
	at kotlin.sequences.FlatteningSequence$iterator$1.ensureItemIterator(Sequences.kt:250)
	at kotlin.sequences.FlatteningSequence$iterator$1.hasNext(Sequences.kt:242)
	at kotlin.sequences.FilteringSequence$iterator$1.calcNext(Sequences.kt:108)
	at kotlin.sequences.FilteringSequence$iterator$1.hasNext(Sequences.kt:133)
	at kotlin.sequences.TransformingSequence$iterator$1.hasNext(Sequences.kt:153)
	at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:647)
	at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:677)
	at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:668)
	at org.kodein.di.internal.KodeinTreeImpl.findBySpecs(KodeinTreeImpl.kt:81)
	at org.kodein.di.internal.KodeinTreeImpl.find(KodeinTreeImpl.kt:99)
	at org.kodein.di.KodeinTree$DefaultImpls.find$default(KodeinTree.kt:32)
	at org.kodein.di.internal.KodeinContainerImpl.factory(KodeinContainerImpl.kt:146)
	at org.kodein.di.KodeinContainer$DefaultImpls.factory$default(KodeinContainer.kt:33)
	at org.kodein.di.KodeinContainer$DefaultImpls.provider(KodeinContainer.kt:80)
	at org.kodein.di.internal.KodeinContainerImpl.provider(KodeinContainerImpl.kt:7)
	at org.kodein.di.KodeinContainer$DefaultImpls.provider$default(KodeinContainer.kt:79)
	at org.kodein.di.internal.DKodeinBaseImpl.Instance(DKodeinImpl.kt:33)
	at org.kodein.di.DKodeinBase$DefaultImpls.Instance$default(DKodein.kt:135)
	at one.core.GrpcScanner.rescan(AccessService.kt:60)
```
